controllers/metallb_controller.go

```diff
@@ -160,91 +161,62 @@ func (r *MetalLBReconciler) SetupWithManager(mgr ctrl.Manager, bgpType string) e
 
-	data.Data["LogLevel"] = metallbv1beta1.LogLevelInfo
-	if config.Spec.LogLevel != "" {
-		data.Data["LogLevel"] = config.Spec.LogLevel
-	}
-
-	var err error
-	data.Data["MetricsPort"], err = valueWithDefault("METRICS_PORT", 7472)
+	actionConfig := new(action.Configuration)
+	chartValueOpts := &values.Options{}
+	chartValues, err := chartValueOpts.MergeValues(getter.All(r.HelmEnvSettings))
 	if err != nil {
 		return err
 	}
-	data.Data["MetricsPortHttps"], err = valueWithDefault("HTTPS_METRICS_PORT", 27472)
+	err = patchPrometheusValues(chartValues, r.Client)
 	if err != nil {
 		return err
 	}
-	data.Data["FRRMetricsPort"], err = valueWithDefault("FRR_METRICS_PORT", 7473)
+	patchControllerValues(chartValues, config, r.PlatformInfo.IsOpenShift())
+	err = patchSpeakerValues(chartValues, config)
 	if err != nil {
 		return err
 	}
-	data.Data["FRRMetricsPortHttps"], err = valueWithDefault("FRR_HTTPS_METRICS_PORT", 27473)
+	// TODO: patch rbac proxy settings if needed.
+	client := action.NewInstall(actionConfig)
+	client.ReleaseName = helmReleaseName
+	client.Namespace = r.HelmEnvSettings.Namespace()
+	client.DryRun = true
+	client.Replace = true
+	client.ClientOnly = true
+	chartPath, err := client.ChartPathOptions.LocateChart(ManifestPath, r.HelmEnvSettings)
 	if err != nil {
 		return err
 	}
-	data.Data["MLBindPort"], err = valueWithDefault("MEMBER_LIST_BIND_PORT", 7946)
+	chart, err := loader.Load(chartPath)
 	if err != nil {
+		logger.Error(err, "metallb chart loading failed")
 		return err
 	}
-
-	objs, err := render.RenderDir(ManifestPath, &data)
+	release, err := client.Run(chart, chartValues)
 	if err != nil {
-		logger.Error(err, "Fail to render config daemon manifests")
+		logger.Error(err, "failed to generate manifests from metallb chart")
 		return err
 	}
-
-	// We shouldn't spam the api server trying to apply PodMonitors if the resource isn't installed.
-	deployPodMonitors := os.Getenv("DEPLOY_PODMONITORS") == "true"
-	if podMonitorAvailable(r.Client) && deployPodMonitors {
-		podmonitors, err := render.RenderDir(PodMonitorsPath, &data)
-		if err != nil {
-			logger.Error(err, "Fail to render PodMonitors manifests")
-			return err
-		}
-		objs = append(objs, podmonitors...)
-	} else {
-		logger.Info("PodMonitors Resource not available in the cluster. Will not try to apply them.")
+	objs, err := render.ParseRenderedManifest(release.Manifest)
+	if err != nil {
+		logger.Error(err, "fail to parse rendered metallb manifests")
+		return err
 	}
 
 	for _, obj := range objs {
-		objNS := obj.GetNamespace()
-		if objNS != "" { // Avoid setting reference on a cluster-scoped resource.
-			if err := controllerutil.SetControllerReference(config, obj, r.Scheme); err != nil {
-				return errors.Wrapf(err, "Failed to set controller reference to %s %s", obj.GetNamespace(), obj.GetName())
-			}
+		objKind := obj.GetKind()
+		// Skip applying role and role binding object
+		if objKind == "Role" || objKind == "RoleBinding" {
```

Comment:
Do we have a way to not import these in the charts?

Reply:
the below logic didn't work on the templated  pod monitor yaml [file](https://github.com/metallb/metallb-operator/blob/756ba4dd10444130a695f9a419ba3ccfd57a490d/bindata/deployment/helm/templates/podmonitor.yaml)
`yq e --inplace '. | select((.kind == "Role" or .kind == "RoleBinding") | not)' "$METALLB_PATH"/charts/metallb/templates/podmonitor.yaml`
and throwing the error
`Error: bad file '_cache/metallb/charts/metallb/templates/podmonitor.yaml': yaml: did not find expected node content`
hence filtered out in the controller code.

Reply:
Ah I see, I thought roles were in a separate yaml but it's not the case for podmonitor.
Let's keep it this way, in the future we may want to split the original helm template in two so filtering is easier.